File: backend/src/models/blog-post.ts
import { Document, model, Schema } from "mongoose";
import { MainUserDocument } from "./main-user";
import MongoPaging from "mongo-cursor-pagination";

export type BlogPostDocument = Document & {
  title: string;
  description: string;
  content: string;
  coverImgURL?: string;
  readTime: number;
  createdAt: Date;
  updatedAt: Date;
  categories: any;
  author: MainUserDocument;
};

const BlogPostSchema = new Schema<BlogPostDocument>(
  {
    title: { type: String, required: true, maxlength: 256, trim: true },
    description: { type: String, required: true, maxlength: 2048, trim: true },
    content: { type: String, required: true, trim: true },
    coverImgURL: { type: String },
    readTime: { type: Number, required: true },
    categories: {
      type: [{ type: Schema.Types.ObjectId, ref: "FoodCategory" }],
      required: true,
    },
    author: { type: Schema.Types.ObjectId, ref: "MainUser", required: true },
  },
  { timestamps: true }
);

BlogPostSchema.plugin(MongoPaging.mongoosePlugin, { name: "paginateBlogPost" });

const BlogPost = model<BlogPostDocument>("BlogPost", BlogPostSchema);
export default BlogPost;

File: flutter-app/lib/models/blog-post/blog-post.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:salt/models/food-category/food-category.dart';
import 'package:salt/models/user/user.dart';

part 'blog-post.g.dart';

@JsonSerializable()
class BlogPost {
  final String id;
  final String title;
  final String description;
  final String content;
  final String coverImgURL;
  final double readTime;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<FoodCategory> categories;
  final User author;

  BlogPost({
    required this.id,
    required this.title,
    required this.description,
    required this.content,
    required this.coverImgURL,
    required this.readTime,
    required this.createdAt,
    required this.updatedAt,
    required this.categories,
    required this.author,
  });

  factory BlogPost.fromJson(Map<String, dynamic> json) {
    return _$BlogPostFromJson(json);
  }

  Map<String, dynamic> toJson() => _$BlogPostToJson(this);
}
